MSPB-215: add support for teammate device

mavriciomart opened the pull request:
5.0 PR: https://github.com/2600hz/monster-ui-voip/pull/406
Ticket: https://2600hz-commercial.atlassian.net/browse/MSPB-215
Screenshots:
<img width="1884" alt="CleanShot 2022-03-17 at 17 40 06@2x" src="https://user-images.githubusercontent.com/33939096/158919505-b0f897da-5065-4925-9319-afa1eb99670c.png">
<img width="1840" alt="CleanShot 2022-03-17 at 17 40 49@2x" src="https://user-images.githubusercontent.com/33939096/158919511-16e27c08-3e2f-4e32-b4c0-f56e2045a47a.png">
<img width="1884" alt="CleanShot 2022-03-17 at 17 41 20@2x" src="https://user-images.githubusercontent.com/33939096/158919513-bf579d67-f605-4de3-8515-69eef40cb1da.png">
<img width="1884" alt="CleanShot 2022-03-17 at 17 42 12@2x" src="https://user-images.githubusercontent.com/33939096/158919521-98a74e4c-d93c-4941-a75f-db9883c99b8f.png">
<img width="1884" alt="CleanShot 2022-03-17 at 17 42 29@2x" src="https://user-images.githubusercontent.com/33939096/158919522-83b91f37-ecf9-4dde-9678-ac621aaa28a1.png">

azefiel commented on submodules/devices/devices.scss:
```diff
@@ -271,8 +272,15 @@
 
-.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > i {
-	font-size: 48px;
+.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > svg {
```
Looks like there is a sizing issue between these changes:
<img width="481" alt="dev_add_device" src="https://user-images.githubusercontent.com/1958193/159091348-8d3dab71-7159-4696-848d-2d4b049b576b.PNG">
and sandbox:
<img width="534" alt="sandbox_add_device" src="https://user-images.githubusercontent.com/1958193/159091350-704952f1-0619-4fa3-beaa-3af98d55e8f9.PNG">

azefiel commented on submodules/devices/devices.scss:
```diff
@@ -271,8 +272,15 @@
+	vertical-align: middle;
+}
+
+.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > svg {
```
Isn't this a duplicate of the rule above? Maybe the last selector was meant to target an `i` tag?

azefiel commented on submodules/devices/views/devices-teammate.html:
```diff
@@ -0,0 +1,182 @@
+				</div>
+			</div>
+
+			{{#if showEmergencyCallerId}}
```
From my understanding of the [spec](https://2600hz-commercial.atlassian.net/wiki/spaces/UCAAS/pages/1747616170/MS+Teams+Billing+Project#E911), this device should not expose e911 configuration at all, unless it changed since writing?

mavriciomart replied:
I left the e911 configuration to follow the mockup from [this section](https://2600hz-commercial.atlassian.net/wiki/spaces/UCAAS/pages/1747616170/MS+Teams+Billing+Project#Adjustment-of-device-options-in-SmartPBX) in the documentation. Will ask for further clarification.

mavriciomart replied:
From a conversation with Bret, we still do want to expose it.

azefiel reviewed (commented):
Looks great overall, minus some concerns regarding defaults definition and other minor typos/possible incorrect config.

It would have been awesome if this changes were broken down into several atomic commits vs one big one, to make the life of reviewers easier 🙏

mavriciomart commented:
@azefiel Thanks for the observations, I already addressed them and tried to separate my work in two commits.

azefiel commented on submodules/devices/devices.js:
```diff
@@ -7,6 +7,10 @@ define(function(require) {
 		return _.every([dest, src], _.isArray) ? src : undefined;
 	};
 
+	var showTeammateDevice = monster.config.hasOwnProperty('extraDevices')
+		&& monster.config.extraDevices.length
+		&& $.inArray('teammate', monster.config.extraDevices) > -1;
+
```
More declarative vs imperative:
```suggestion
	var showTeammateDevice = _
	    .chain(monster.config)
	    .get('extraDevices', [])
	    .includes('teammate')
	    .value();

```

azefiel commented on submodules/users/users.js:
```diff
@@ -3763,7 +3767,8 @@ define(function(require) {
 				assignedDevices: _.keyBy(assigned, 'id'),
-				unassignedDevices: _.keyBy(unassigned, 'id')
+				unassignedDevices: _.keyBy(unassigned, 'id'),
+				showTeammateDevice: showTeammateDevice
```
If you think about it, that new prop is not related to this function, in fact only the template cares about it.

How about inserting it as a template prop directly?

```js
template = $(self.getTemplate({
	name: 'devices',
	data: _,merge({
            showTeammateDevices: showTeammateDevices
        }, formattedResults),
	submodule: 'users'
}));
```

azefiel commented on submodules/devices/devices.js:
```diff
@@ -58,7 +62,7 @@ define(function(require) {
 					'ata',
 					'sip_uri',
-					'teammate'
+					showTeammateDevice ? 'teammate' : undefined
```
This will potentially create an empty element, which in the grand scheme of things might not be a big deal BUT it would be nice to keep our data structures clean to avoid unexpected behavior down the road.

How about flattening nested arrays instead?

```js
addableDeviceTypes: _.flatten([[
      'sip_device',
      'cellphone',
      'smartphone',
      'softphone',
      'landline',
      'fax',
      'ata',
      'sip_uri'
  ],
  cond ? [
      'teammate'
  ] : []
])

azefiel reviewed (commented):
Not sure what expectations are but, should devices of types listed in `extraDevices` be selectable/listable in the UI when that list is empty?

For example, `extraDevices=['teammate']` and I create a teammate device. Later, `extraDevices` gets updated and `teammate` removed. Should the UI still list those devices in the UI when fetching device documents?

mavriciomart commented:
Thanks @azefiel! I addressed your comments in a newer commit.

> Not sure what expectations are but, should devices of types listed in extraDevices be selectable/listable in the UI when that list is empty?
For example, extraDevices=['teammate'] and I create a teammate device. Later, extraDevices gets updated and teammate removed. Should the UI still list those devices in the UI when fetching device documents?

The idea of this new key is that it will be set once the TeamMate integration has been configured for the cluster, and will allow the user to create a new teammate device.

I also updated the key name to `allowedExtraDeviceTypes` to be more clear of its purpose.

mavriciomart commented:
@azefiel I got clarification from Karl on the above question.

> If they don't have teammate enabled but the system has devices of that type I would assume they must have enabled the device type at some point.  that means if they simply disappear they would have no way to remove them which means they have creds floating around they cant reset/remove (fraud / security risk) and no way to delete to avoid charges (billing).  So, yup the new parameter should only control showing the create option.

azefiel commented:
Thanks for getting clarification on this @mavriciomart , that explanation is sound and makes total sense in hindsight.

harenson commented on submodules/devices/devices.js:
```diff
@@ -1218,7 +1230,19 @@ define(function(require) {
+						media: {
+							encryption: {
+								enforce_security: true,
+								webrtc: false,
```
`webrtc` should go inside `media` not inside `media.encryption`.
